Finish the state transition proof verifier revamp

liuchengxu opened the pull request:
This is is the last pice of series of the state transition proof verifier revamp, the main point of this PR is to integrate the `domain-block-preprocessor` component on the verifier side. There will be probabaly some more tests close to the production work flow, but the core features as mentioned the tracking issue #1230 are complete after this PR. 

Once #1342 is merged into the main branch, I'd love to have another look at the basic tests in the state transition aspect, and then we unblock #1020 to re-enable fraud-proof again.

Close #1274
Close #1230

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

liuchengxu committed:
Retrieve state transition data for InitializeBlock

liuchengxu committed:
Introduce CoreDomainExtrinsicsBuilder

liuchengxu committed:
Enable fraud proof tests

NingLin-P commented on crates/subspace-fraud-proof/src/invalid_state_transition_proof.rs:
```diff
@@ -449,7 +449,26 @@ where
+                    Default::default(),
+                    Default::default(),
+                    parent_hash,
+                    Default::default(),
```
We may need the `primary_block_info` digest here.

liuchengxu replied:
Thanks for the comment, yes, I did forget to add a TODO, rebased and pushed a commit to add it. The code changes to resolve it won't be significant, but we need to add a new test to cover the regular fraud proof creation and verification route. I expect you to take that, but I could also work on it once I process the other priorities.

vedhavyas replied:
are we not handling the proper digests in this PR ?

NingLin-P replied:
> but we need to add a new test to cover the regular fraud proof creation and verification route. I expect you to take that, but I could also work on it once I process the other priorities.

Sure, it is on my plan for this week.

liuchengxu replied:
>are we not handling the proper digests in this PR ?

Nope. We found the system domain digests issue on the proof creation as described [here](https://github.com/subspace/subspace/pull/1344/commits/b5dca024e92e27eb98377ab0eb0ac3a060408f5c#diff-77459f83c5cf59cea043e7ca65b282ab5ff94c1ac8b463dda1588cdef8ba7d80R675) before and agreed @NingLin-P will take care of it with a test added. Since the creation and verification are basically in one single test flow, I'll leave it as is to Ning.

liuchengxu replied:
The digest issue of system domain has been addressed in the latest commit, I had to fix it right now to pass the test.

NingLin-P replied:
The fraud proof test is also ready here https://github.com/subspace/subspace/tree/fraud-proof-test, I will open a PR for it after this PR is merged.

liuchengxu committed:
Introduce BuildDomainExtrinsics trait and SystemDomainExtrinsicsBuilder

`primary_hash` is added to `ExecutionPhase::ApplyExtrinsic` as it's
needed to retrieve the original primary block.

liuchengxu committed:
Add a TODO for the state transition data for system domain InitializeBlock

The code changes for the fix would be minor, the main workload is writing
a test to cover the workflow of fraud proof creation and verification in the
production environment so that we can ensure the same glitch won't
happen again.

vedhavyas commented on crates/sp-domains/src/fraud_proof.rs:
```diff
@@ -16,7 +16,10 @@ pub enum ExecutionPhase {
-    ApplyExtrinsic(u32),
+    ApplyExtrinsic {
+        extrinsic_index: u32,
+        primary_hash: H256,
```
i'm not sure about this one TBH. The verifier should have this info already. If not they should be able to get this from the receipts on the state through `primary_hash` no?

liuchengxu replied:
We need to explicitly declare which primary hash the proof is corresponding to considering the forks. The verifier only knows about `primary_parent_hash` not `primary_hash`. There are potentially multiple receipts at a certain height, which one are you referring to that can get `primary_hash` specifically corresponding to this proof?

NingLin-P replied:
This reminds me that we may need to verify the integrity of `primary_parent_hash` and `primary_hash` to ensure `primary_parent_hash` is indeed `primary_hash`'s parent hash.

Otherwise, since `primary_parent_hash` is used to fetch the runtime code, the attacker can set it to an old value and construct a state transition proof that is valid for the old runtime but invalid for the runtime that the receipt targetted and still pass the verification.

vedhavyas replied:
> There are potentially multiple receipts at a certain height,

True but there will only one canonical primary block and this can be retrieved from the embed primary client or through `PrimaryBlockHash` on the pallet-receipts since we already know the block number.

vedhavyas replied:
> This reminds me that we may need to verify the integrity of primary_parent_hash and primary_hash to ensure primary_parent_hash is indeed primary_hash's parent hash.

yes or we can just extract this info from the client and verification is done implicitly.

liuchengxu replied:
Using the client needs to traverse back from the tip to `primary_parent_hash` in order to find the canonical `primary_hash`, there is a concern that an attacker can forge a proof that sets `primary_parent_hash` to one million blocks ago, since this verification is done before the normal `validate_transaction`, then it will do one million lookups which are unacceptable, some more work is necessary to prevent it.

Retrieving it from `PrimaryBlockHash` looks good to me so that we don't have to do more verification, this requires a new interface in `ReceiptsApi`, it looks like I also have a chance to address this backlog https://github.com/subspace/subspace/pull/1240#issuecomment-1476212007 by refactoring the state root verifier.

liuchengxu replied:
It's now retrieved from `PrimaryBlockHash` in the lastest push, but `pre_state_root` remains unchanged as some non-trivial test fixes are needed to refactor it, a TODO is added which could be probably resolved when the fraud proof tests are improved.

vedhavyas reviewed (commented):
Overall make sense with integration but unsure about the primary hash

NingLin-P reviewed (commented):
Rest LGTM

liuchengxu committed:
Add TODO for refactoring `pre_state_root` and remove unnecessary trait bound in runtime_api_full

liuchengxu committed:
Retrieve primary_hash from `PrimaryBlockHash` in pallet-receipts

Added `primary_hash()` to the trait based on runtime api `ReceiptsApi`.

`H256` is chosen as the return value type in order to avoid the hassle
of different Block types. The runtime api call returns `ParentChainBlock`
type and then needed to be converted into `PBlock` type. Adding a
generic to trait `VerifyInvalidStateTransitionProof` makes the code run
into the generics rabbit hole IMO.

liuchengxu committed:
Move test-specific code into tests and fix fraud proof test

vedhavyas commented on crates/subspace-fraud-proof/src/invalid_state_transition_proof.rs:
```diff
@@ -240,6 +257,13 @@ pub trait VerifyPrePostStateRoot {
     ) -> Result<(), VerificationError>;
+
+    /// Returns the hash of primary block at height `domain_block_number`.
+    fn primary_hash(
```
Adding the func to this trait seems weird, Maybe a different trait would be nicer

liuchengxu replied:
Agree. I left it as is intentionally for simplicity as we'll probably rename the trait `VerifyPrePostStateRoot` to be something more appropriate (I don't yet have a good idea about this :P) when refactoring `pre_state_root`, and then another trait is unnecessary.

vedhavyas reviewed (commented):
Make sense!

liuchengxu committed:
Merge branch 'main' into finish-state-transition-proof-verifier-revamp

liuchengxu committed:
Fix test `execution_proof_creation_and_verification_should_work`

liuchengxu committed:
Fix call data of ExecutionPhase::InitializeBlock and test

- This commit fixes a bug in `Header::new()` when creating the new header:
`parent_number` -> `primary_number`.
- Resolves https://github.com/subspace/subspace/pull/1344#discussion_r1156848535
